src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioPortableExecutableReference.FileTimeStampProvider.cs

```diff
@@ -0,0 +1,114 @@
+
+internal partial class VisualStudioMetadataReferenceManager
+{
+    private sealed partial class VisualStudioPortableExecutableReference
+    {
+        /// <summary>
+        /// Provides a performant mechanism to get the timestamp of a file. Caches the timestamp of requested
+        /// files and watches the directories containing those files for changes.
+        /// </summary>
+        private sealed class FileTimeStampProvider
+        {
+            private static FileTimeStampProvider? s_instance;
+
+            private readonly Dictionary<string, DateTime> _cachedTimestamps = [];
+            private readonly Dictionary<string, IFileChangeContext> _fileChangeContexts = [];
+
+            private FileTimeStampProvider(Workspace workspace)
+            {
+                _ = workspace.RegisterWorkspaceChangedHandler(OnWorkspaceChanged);
+            }
+
+            private void OnWorkspaceChanged(WorkspaceChangeEventArgs args)
+            {
+                switch (args.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                        ResetCaches();
+                        break;
+                }
+
+                return;
+
+                void ResetCaches()
+                {
+                    lock (_cachedTimestamps)
+                    {
+                        foreach (var (_, context) in _fileChangeContexts)
+                            context.FileChanged -= OnContextFileChanged;
+
+                        _fileChangeContexts.Clear();
+                        _cachedTimestamps.Clear();
+                    }
+                }
+            }
+
+            public static DateTime GetTimeStamp(string fullPath, IFileChangeWatcher watcher, Workspace workspace)
+               => GetInstance(workspace).GetTimeStamp(fullPath, watcher);
+
+            private static FileTimeStampProvider GetInstance(Workspace workspace)
+            {
+                return s_instance ??= new FileTimeStampProvider(workspace);
+            }
+
+            private DateTime GetTimeStamp(string fullPath, IFileChangeWatcher watcher)
+            {
+                DateTime timestamp;
+
+                lock (_cachedTimestamps)
+                {
+                    // Attempt to use the cached timestamp for this file. 
+                    if (_cachedTimestamps.TryGetValue(fullPath, out timestamp))
+                        return timestamp;
+                }
+
+                // If we don't have a cached timestamp, we'll need to recalculate. Do this outside
+                // the lock as it can take some time.
+                var directory = Path.GetDirectoryName(fullPath);
+                timestamp = FileUtilities.GetFileTimeStamp(fullPath);
+
+                lock (_cachedTimestamps)
```

Comment:
There is a race here where if there is a file write between the read of the time stamp and taking the lock that things could get out of order:

- Thread A reads the older timestamp. Now mentally pause this thread.
- File is touched.
- Thread B reads the newer timestamp. Takes lock, updates the map.
- Mentally unpause Thread A. Writes the old timestamp.

You'll have to think if this matters or not.